vowpalwabbit/core/src/confidence_sequence_robust.cc

```diff
@@ -0,0 +1,403 @@
+// Copyright (c) by respective owners including Yahoo!, Microsoft, and
+// individual contributors. All rights reserved. Released under a BSD (revised)
+// license as described in the file LICENSE.
+
+#include "vw/core/confidence_sequence_robust.h"
+
+#include "vw/core/model_utils.h"
+
+#include <cassert>
+#include <cmath>
+
+namespace VW
+{
+g_tilde::g_tilde(double k) : k(k), log_k(std::log(k)), sum_x(0.0), sum_low_v(0.0), sum_mid_v(0.0), t(0)
+{
+  assert(1 < k && k < 2);
+}
+
+double g_tilde::histo_variance(double lam_sqrt_tp1) const
+{
+  double sqrt_tp1 = std::sqrt(t + 1);
+  double ret_val = 0.0;
+  for (const std::pair<const std::pair<uint64_t, bool>, double>& hist_p : sum_v_histo)
+  {
+    uint64_t n = hist_p.first.first;
+    bool strong_term = hist_p.first.second;
+    double c = hist_p.second;
+    double x_raw = std::pow(k, n);
+    double x_ds = x_raw / sqrt_tp1;
+    double curr_val = 0.0;
+    if (strong_term) { curr_val = std::pow(lam_sqrt_tp1 * (k - 1) * x_ds / (1 + lam_sqrt_tp1 * k * x_ds), 2); }
+    else { curr_val = lam_sqrt_tp1 * x_ds - std::log1p(lam_sqrt_tp1 * x_ds); }
+    ret_val += c * curr_val;
+  }
+  return ret_val;
+}
+
+void g_tilde::histo_insert(double x)
+{
+  uint64_t n = std::floor(std::log(x) / log_k);
+  double x1 = std::pow(k, n);
+  double alpha = (k * x1 - x) / ((k - 1.0) * x1);
+  // TODO: Kahan summation
```

Comment:
is this comment stale?

Reply:
It's copied from Paul's Python code, not sure if its still needed